# JCPyV_transcriptome
![JCPyV_genome_github](https://github.com/user-attachments/assets/27a00529-6e00-496d-b631-444e2323d1a1)
The R code shared here was used for the analysis of JC polyomavirus (JCPyV) transcriptome data. This repository also contains a portion of the data analyzed in the paper.

|Directory|Contents|
|------|------|
|R|R code for the analysis of transcriptome data.|
|data|Sequencing data in CSV or FASTA format.|

# Code
1. Fig3_JCPyV_transcripts.r<br>
   For the visualization of JCPyV transcripts identified in the study.<br>
2. Fig4_Fig5_bubble_chart.r<br>
   For the visualization of read counts obtained in JCPyV wraparound amplicon sequencing using a bubble chart.<br>
3. Fig6_JCPyV_NCCR.r<br>
   For the visualization of the noncoding control region (NCCR) of the JCPyV genome.<br>

# Data
1. Fig3<br>
   Contains a csv file summarizing the position of splicing acceptors/donors.<br>
2. Fig4_Fig5<br>
   Contains zip-archived spans files derived from amplicon sequencing of wraparound transcripts. Csv files summarize properties of JCPyV transcripts identified in this study.
4. Fig6<br>
   Contains zip-archived fasta files of the JCPyV genome and a csv file summarizing rearrangements/insertions/deletions in NCCR.
